sympy/geometry/point.py

```diff
@@ -234,6 +256,95 @@ def ambient_dimension(self):
+            return Point([1] + (dim - 1)*[0])
+        if self[1] == S.Zero:
+            return Point([0,1] + (dim - 2)*[0])
+        # if the first two coordinates aren't zero, we can create a non-zero
+        # orthogonal vector by swapping them, negating one, and padding with zeros
+        return Point([-self[1], self[0]] + (dim - 2)*[0])
+
+    @staticmethod
+    def affine_rank(*points):
+        """The affine rank of a set of points is the dimension
+        of the smallest affine space containing all the points.
+        For example, if the points lie on a line (and are not all
+        the same) their affine rank is 1.  If the points lie on a plane
+        but not a line, their affine rank is 2."""
+
+        if len(points) == 0:
+            raise TypeError("Must have at least one point")
+
+        # make sure we're genuinely points
+        # and translate to every point to the origin
+        # XXX: issue #11238 we need to pre-simplify otherwise
+        # the rank calculation will be incorrect
+        origin = simplify(Point.pointify(points[0]))
+        points = [Point.pointify(p, dimension=origin.ambient_dimension) - origin for p in points[1:]]
+
+        m = Matrix([p.args for p in points])
+        # XXX: issue #9480 we need `simplify=True` otherwise the
+        # rank may be computed incorrectly
+        return m.rank(simplify=True)
+
+    @staticmethod
+    def project(a, b):
+        """Project the point `a` onto the non-zero point `b`."""
+        a, b = Point.pointify(a), Point.pointify(b)
+        if b.is_zero:
+            raise ValueError("Cannot project to the zero vector.")
+        if a.ambient_dimension != b.ambient_dimension:
+            raise ValueError("Points must be of the same dimension")
+
+        return b*(a.dot(b) / b.dot(b))
+
+    @staticmethod
+    def pointify(a, dimension=None, ensure_geometry_entity=True, ensure_point=True, **kwargs):
+        """If `a` is a `Point` or a non-iterable, it is left alone.
+        Otherwise, it is converted to a point with `kwargs` passed along
+        to `Point()`. If `dimension` is set (default `None`), a dimension check is carried out.
+        If `ensure_geometry_entity` is set (default `True`), a type check for GeometryEntity
+        is carried out and the return value will be a GeometryEntity.
+        If `ensure_point` is set (default `True`), a type check for Point is carried out
+        and the return value will be a Point."""
+
+        if isinstance(a, Point):
+            if dimension and a.ambient_dimension != dimension:
+                raise ValueError("Point {} must be of dimension {}.".format(a, dimension))
+            return a
+
+        # any interable that is traversed in a canonical way
+        # (which excludes things like FiniteSets) can be turned
+        # into a point.  This check can be made more general, but for now
+        # exclude known bad objects
+        if iterable(a) and not isinstance(a, (FiniteSet, set, dict)):
```

Comment:
I avoided `is_sequence` so that `Point(x**2 for x in [1,2,3])` would work.  That is, `g=(x**2 for x in [1,2,3])` passes the `pointify` tests, but `is_sequence(g) == False`.